swift/WorkflowTesting/Sources/RenderExpectations.swift

```diff
@@ -99,6 +104,24 @@ public struct ExpectedWorker {
     }
 }
 
+public struct ExpectedSideEffect<WorkflowType: Workflow> {
+    let key: AnyHashable
+    let action: ((RenderContext<WorkflowType>) -> Void)?
```

Comment:
What's the purpose of this closure?

Reply:
> Currently, the only init takes in key and action, however, I wanted to make sure we can pass in arbitrary block of code in the future to ensure it's in line with the `runSideEffect` api.

Reply:
But what is the block intended to do? If you're writing a test for a workflow that directly invokes `runSideEffect`, what kinds of things would you actually want to assert on?

1. That `runSideEffect` was called with a given key.
2. That `runSideEffect` was _not_ called with a given key.
3. Given that a side effect was ran with a given key, that the side effect lambda actually performed some side effect (e.g. called `logger.log()` with some message).
4. Given that a side effect was ran with a given key, that the side effect lambda registers the correct logic to be executed in the `onEnded` handler.
5. Given that a side effect was ran with a given key, that the side effect lambda sent some action to the event sink.

1 can definitely be done with this API, presumably 2 can as well (by _not_ expecting the key?). I'm not sure how close the Swift render testing API is to the kotlin one yet, but if the testing infrastructure runs the action immediately, then 3 and 5 can be done with existing APIs (3 by asserting the side effect fake was invoked after the render call, 5 using [`verifyAction(Result)`](https://square.github.io/workflow/kotlin/api/workflow/com.squareup.workflow.testing/-render-test-result/)). The only special case is 4, which requires an API to end the lifetime. Some ways to do this:

1. Pass a flag in `ExpectedSideEffect` that, if set, will cause the `onEnded` handlers to be invoked immediately.
2. An additional API on [`RenderTestResult`](https://square.github.io/workflow/kotlin/api/workflow/com.squareup.workflow.testing/-render-test-result/) called something like `endSideEffect(key: String): RenderTestResult` that executes the end actions for the given side effect and returns a `RenderTestResult` that allows verifying if any actions were sent to the sink in the `onEnded` handler.

I think 1 is the simpler option, although both could be tricky (if the side effect immediately sends an action, and then sends another action when ending, the initial action will always be evaluated first and a second render pass would be needed to evaluate the second, and that's hard to express in the render tester API since it, by definition, tests a single render pass). Perhaps we don't actually need to need to provide the ability to test side effect implementations in this API, and can use the more general "integration" testing API?

In any case, this lambda parameter doesn't seem to address any of these use cases so I'm also not sure what purpose it serves.

Reply:
This is in response to @bencochran 's feedback here: https://github.com/square/workflow/pull/1174#discussion_r430828163

`ExpectedWorker` allows you to send an `Output` back to the `RenderContext` when executed. Trying to provide the same ability here.

The `action` is actually not public, it's not meant to be used directly. The two public initializers are:

```swift
public init(key: AnyHashable)
public init<ActionType: WorkflowAction>(key: AnyHashable, action: ActionType) where ActionType.WorkflowType == WorkflowType
```

I wanted to make sure `ExpectedSideEffect` is implemented similarly to how `SideEffects` are implemented and hence in the future we can provide additional ways to pass `WorkflowAction`s to the `RenderContext`.

Reply:
I'm worried we never quite converged our testing APIs, and now they're diverging more. Need to dig into this more to understand where the differences are. I could have sworn we had an issue to track testing APIs but I can't find it.

Reply:
> I could have sworn we had an issue to track testing APIs but I can't find it.

#702?

Reply:
I think the public parts of `ExpectedSideEffect` as written here fits with the overall current shape of our `RenderTester`. Though, yeah, I’d prefer we converged in the direction of #702 and what Kotlin has.

I’m not sure of the need internally to hold a `((RenderContext<WorkflowType>) -> Void)?` vs an `AnyWorkflowAction<WorkflowType>?` but that’s implementation detail that I’m less concerned with, I think?

Reply:
My plan is to work on the Kotlin side of this once I've split the repo out, this week. If yall are happy with this then SGTM, I'll not look too hard, see what I come up with in Kotlin, hopefully it's the same! And if not, I'll be in a much better position to debate 😉